src/analyses/ai_domain.h

```diff
@@ -0,0 +1,123 @@
+/*******************************************************************\
+
+Module: Abstract Interpretation
+
+Author: Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
+/// \file
+/// Abstract Interpretation Domain
+
+#ifndef CPROVER_ANALYSES_AI_DOMAIN_H
+#define CPROVER_ANALYSES_AI_DOMAIN_H
+
+#include <util/expr.h>
+#include <util/json.h>
+#include <util/make_unique.h>
+#include <util/xml.h>
+
+#include <goto-programs/goto_model.h>
+
+// forward reference the abstract interpreter interface
+class ai_baset;
+
+/// The interface offered by a domain, allows code to manipulate domains without
+/// knowing their exact type.
+// don't use me -- I am just a base class
+// please derive from me
```

Comment:
Feel free to disregard as out of scope, but rather than pleading, why not just make the constructor protected? In fact - I suspect no need at all for this comment since there is a pure virtual function here anyway.